### PR TITLE
[RLlib; Offline RL] Make data pipeline better configurable and tuneable for users.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

The new `Offline RL API` makes direct use of  `ray.data.Datasets` and there `map_batches` and `iter_batches`  methods to transform and iterate over batches. These methods allow for configuration to tune the data pipeline. This PR proposes a way for users  to easily pass arguments to these methods in the ´AlgorithmConfig` to tune their data pipelines. 

In addition the `OfflinePreLearner` is moved to its own file from which users can override its behavior in its `_map_episodes` and `__call__`  method to define in detail how their data should be transferred to a batch/episodes.

Furthermore, some bug fixes have been made.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/algorithm_config.py [hunk 1]
```diff
@@ -2415,6 +2423,27 @@ def offline_data(
...
                 needed.
+            map_batches_kwargs: `kwargs` for the `map_batches` method. These will be
+                passed into the `ray.data.Dataset.map_batches` method when sampling
+                without checking. If no arguments passed in the default arguments `{
+                'concurrency': max(2, num_learners), 'zero_copy_batch': True}` is
+                used. Use these `kwargs`` together with the `input_read_method_kwargs`
+                and `iter_batches_kwargs` to tune the performance of the data pipeline.
+            iter_batches_kwargs: `kwargs` for the `iter_batches` method. These will be
+                passed into the `ray.data.Dataset.iter_batches` method when sampling
+                without checking. If no arguments passed in the default arguments `{
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:2434]: nit: "argument" (singular)

Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:2434]:
```suggestion
                without checking. If no arguments are passed in, the default argument `{
```

File: rllib/algorithms/algorithm_config.py [hunk 2]
```diff
@@ -2415,6 +2423,27 @@ def offline_data(
                 schema used is `ray.rllib.offline.offline_data.SCHEMA`. If your data set
                 contains already the names in this schema, no `input_read_schema` is
                 needed.
+            map_batches_kwargs: `kwargs` for the `map_batches` method. These will be
+                passed into the `ray.data.Dataset.map_batches` method when sampling
+                without checking. If no arguments passed in the default arguments `{
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:2428]:
```suggestion
                without checking. If no arguments are passed in, the default argument `{
```

File: rllib/algorithms/algorithm_config.py [hunk 3]
```diff
@@ -2470,6 +2499,8 @@ def offline_data(
             self.input_read_method_kwargs = input_read_method_kwargs
         if input_read_schema is not NotProvided:
             self.input_read_schema = input_read_schema
+        if prelearner_class is not NotProvided:
+            self.prelearner_class = prelearner_class
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:2503]: Wait, where do these get assigned?
`input_read_method_kwargs`
`map_batches_kwargs`

File: rllib/offline/offline_data.py [hunk 4]
```diff
@@ -151,200 +131,16 @@ def default_read_method_kwargs(self):
...
         }
 
+    @property
+    def default_map_batches_kwargs(self):
+        return {
+            "concurrency": max(2, self.config.num_learners),
+            "zero_copy_batch": True,
+        }
 
-class OfflinePreLearner:
```
Comment (sven1977) [inline rllib/offline/offline_data.py:155]: Thanks for separating these!

File: rllib/tuned_examples/bc/cartpole_bc.py [hunk 5]
```diff
@@ -49,7 +49,12 @@
     # as remote learners.
     .offline_data(
         input_=[data_path],
-        input_read_method_kwargs={"override_num_blocks": max(args.num_gpus, 1)},
+        input_read_method_kwargs={"override_num_blocks": max(args.num_gpus, 2)},
```
Comment (sven1977) [inline rllib/tuned_examples/bc/cartpole_bc.py:52]: Can we add a one-line comment to each of these settings, explaining, why we have to set these and what they do under the hood?

File: rllib/offline/offline_prelearner.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    # TODO (simon): Add remove as soon as we are new stack only.
+    "agent_index": "agent_index",
+    "dones": "dones",
+    "unroll_id": "unroll_id",
+}
+
+
+@ExperimentalAPI
+class OfflinePreLearner:
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
```
Comment (sven1977) [inline rllib/offline/offline_prelearner.py:48]: Can we add a big docstr here to this class that explains what it does, when it's used, how it works, etc..

File: rllib/algorithms/algorithm_config.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/algorithms/algorithm_config.py:2503]: Up in the file. Where all attributes get default values.

Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:2503]: No, sure, but if the user sets them through the `offline_data()` method, currently nothing would happen, correct?

Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:2503]: Ah, sorry, didn't see this. I think b/c it hadn't been changed in this PR. All good.

**Review by sven1977 — APPROVED** [review]
Looks good to me now! Thanks @simonsays1980 !! :)

**Pull request merged**